onnxruntime/core/providers/coreml/builders/impl/squeeze_op_builder.cc

```diff
@@ -54,38 +54,95 @@ void SqueezeOpBuilder::AddInitializersToSkip(ModelBuilder& model_builder, const
+      std::sort(axes.begin(), axes.end());
+      // For example: Given an input tensor (data) of shape [3, 4, 5],
+      // then Unsqueeze(data, axes=[0, 4]) outputs a tensor (expanded) containing same data as data but with shape [1, 3, 4, 5, 1].
+      for (size_t i = 0, ori_i = 0, axes_i = 0; i < new_shape.size(); i++) {
+        if ((axes_i >= axes.size() || static_cast<int64_t>(i) != axes[axes_i]) && input_shape.size() >= ori_i) {
+          new_shape[i] = input_shape[ori_i++];
+        } else {
+          axes_i++;
+        }
+      }
+      AddOperationInput(*op, "shape", model_builder.AddConstant(op->type(), "shape", new_shape));
+    }
+    AddOperationOutput(*op, *node.OutputDefs()[0]);
+    model_builder.AddOperation(std::move(op));
+  } else  // NOLINT
+#endif
+  {
+    if (axes.empty()) {
+      coreml_squeeze->set_squeezeall(true);
+    } else {
+      *coreml_squeeze->mutable_axes() = {axes.cbegin(), axes.cend()};
+      coreml_squeeze->set_squeezeall(false);
+    }
 
-  *layer->mutable_input()->Add() = node.InputDefs()[0]->Name();
-  *layer->mutable_output()->Add() = node.OutputDefs()[0]->Name();
+    *layer->mutable_input()->Add() = node.InputDefs()[0]->Name();
+    *layer->mutable_output()->Add() = node.OutputDefs()[0]->Name();
 
-  model_builder.AddLayer(std::move(layer));
+    model_builder.AddLayer(std::move(layer));
+  }
   return Status::OK();
 }
 
 bool SqueezeOpBuilder::IsOpSupportedImpl(const Node& node, const OpBuilderInputParams& input_params,
-                                         const logging::Logger& /*logger*/) const {
+                                         const logging::Logger& logger) const {
   // Squeeze opset 13 uses input 1 as axes, if we have input 1 then it needs to be an initializer
-  const auto& initializers = input_params.graph_viewer.GetAllInitializedTensors();
-  if (node.SinceVersion() > 12 && node.InputDefs().size() > 1) {
-    const auto& axes_name = node.InputDefs()[1]->Name();
-    if (!Contains(initializers, axes_name)) {
-      LOGS_DEFAULT(VERBOSE) << "Input axes of Squeeze must be known";
+  const auto& input_defs = node.InputDefs();
+  if (node.SinceVersion() > 12 && input_defs.size() > 1) {
+    const auto& axes_name = input_defs[1]->Name();
+    if (!input_params.graph_viewer.GetConstantInitializer(axes_name)) {
+      LOGS(logger, VERBOSE) << "Input axes of Squeeze must be known";
       return false;
     }
   }
 
+  if (node.OpType() == "Unsqueeze") {
+    if (!input_params.create_mlprogram) {
+      return false;
+    }
+    int64_t rank = -1;
+    if (node.SinceVersion() > 12) {
+      const auto& axes_tensor = *input_params.graph_viewer.GetConstantInitializer(input_defs[1]->Name());
+      Initializer unpacked_tensor(axes_tensor);
+      rank = unpacked_tensor.size();
+    } else {
+      NodeAttrHelper helper(node);
+      auto axes = helper.Get("axes", std::vector<int64_t>());
+      rank = static_cast<int64_t>(axes.size());
+    }
+    std::vector<int64_t> input_shape;
```

Comment:
nit: code is (IMHO) a lot easier to read when different blocks aren't joined together as the whitespace provides a quickly parsed visual indicator the next line is not attached to the previous block.

```suggestion
    if (!input_params.create_mlprogram) {
      return false;
    }

    int64_t rank = -1;
    if (node.SinceVersion() > 12) {
      const auto& axes_tensor = *input_params.graph_viewer.GetConstantInitializer(input_defs[1]->Name());
      Initializer unpacked_tensor(axes_tensor);
      rank = unpacked_tensor.size();
    } else {
      NodeAttrHelper helper(node);
      auto axes = helper.Get("axes", std::vector<int64_t>());
      rank = static_cast<int64_t>(axes.size());
    }
    
    std::vector<int64_t> input_shape;
```